bson/codec_options.py

```diff
@@ -32,10 +33,63 @@ def _raw_document_class(document_class):
     return marker == _RAW_BSON_DOCUMENT_MARKER
 
 
+class TypeCodecBase(object):
+    @property
+    def python_type(self):
+        """The Python type to be converted into something serializable."""
+        raise NotImplementedError
+
+    @property
+    def bson_type(self):
+        """The BSON type to be converted into our own type."""
+        raise NotImplementedError
+
+    def transform_bson(self, value):
+        """Convert the given BSON value into our own type."""
+        raise NotImplementedError
+
+    def transform_python(self, value):
+        """Convert the given Python object into something serializable."""
+        raise NotImplementedError
+
+
+class TypeRegistry(object):
+    def __init__(self, *type_codecs):
+        self.__args = type_codecs
+        self._encoder_map = {}
+        self._decoder_map = {}
+        for codec in type_codecs:
+            if not isinstance(codec, TypeCodecBase):
+                raise TypeError("Expected instance of %s, got %s instead" % (
+                    TypeCodecBase.__name__, type(codec).__name__))
+            try:
+                python_type = codec.python_type
+            except NotImplementedError:
+                pass
+            else:
+                self._encoder_map[python_type] = codec.transform_python
+
+            try:
+                bson_type = codec.bson_type
+            except NotImplementedError:
+                pass
+            else:
+                self._decoder_map[bson_type] = codec.transform_bson
+
+    def __repr__(self):
+        return '%s%r' % (self.__class__.__name__, self.__args)
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return False
```

Comment:
I think you want to return NotImplemented here. https://docs.python.org/3/library/constants.html#NotImplemented

Reply:
Ah, good point.